### PR TITLE
Classification

### PR DESCRIPTION
**Pull request by Eli-Persky**
Added classification code.
Altered method of sampling nll as discussed.

### CODE REVIEW

File: classification/class_task.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            out = torch.nn.Softmax(dim=1)(self.net(X))
+            probs = probs + out / self.test_samples
+        preds = torch.argmax(probs, dim=1)
+        return preds
+
+    def evaluate(self, test_loader):
+        print('Evaluating on validation data')
+        correct = 0
+        total = 0
+
```
Comment (tennisonliu) [inline classification/class_task.py:76]:
```suggestion
self.net.eval()
```

File: classification/class_task.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            'hidden_units': self.hidden_units,
+            'mode': self.mode,
+            'mu_init': self.mu_init,
+            'rho_init': self.rho_init,
+            'prior_init': self.prior_init,
+            'mixture_prior': self.mixture_prior
+        }
+        self.net = BayesianNetwork(model_params).to(DEVICE)
+        self.optimiser = torch.optim.Adam(self.net.parameters(), lr=self.lr)
+        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimiser, step_size=5000, gamma=0.5)
```
Comment (tennisonliu) [inline classification/class_task.py:47]:
```suggestion
        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimiser, step_size=100, gamma=0.5)
```
If we want to use scheduler for classification, should decrease step_size since only training for 300 epochs.

File: .gitignore [hunk 3]
```diff
@@ -2,4 +2,5 @@ __pycache__/
 */__pycache__/
 /data
 /runs
-poc.py
```
Comment (tennisonliu) [inline .gitignore:5]: can we stop tracking __pycache files, see link below on clearing git cache. https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore

File: networks.py [hunk 4]
```diff
@@ -124,26 +125,30 @@ def get_nll(self, outputs, target, sigma=1.):
         if self.mode == 'regression':
             ll = torch.distributions.Normal(outputs, sigma).log_prob(target).sum()
         elif self.mode == 'classification':
-            ll = nn.CrossEntropy()
+            ll = -nn.CrossEntropyLoss()(outputs, target)
```
Comment (tennisonliu) [inline networks.py:128]: are we sure this should be negative? on line 131, the code is returning -ll

File: .gitignore (hunk 3 above)
Comment (tennisonliu) [reply .gitignore:5]: also can we stop tracking /runs. I suspect you have to clear cache, see link attached above.

**Review by tennisonliu — CHANGES_REQUESTED** [review]
Looks good, added a few minor comments. One thing to make sure about is the negative cross entropy loss.

File: networks.py (hunk 4 above)
Comment (Eli-Persky) [reply networks.py:128]: Yes because CrossEntropyLoss has a negative built in.

Comment (tennisonliu) [reply networks.py]: Agreed.

File: .gitignore (hunk 3 above)
Comment (tennisonliu) [reply .gitignore:5]: Resolved.

**Review by tennisonliu — APPROVED** [review]
Good stuff.
